### ISSUE TITLE
autosummary cannot handle umlauts in function names on Windows.

### ISSUE THREAD

**Issue by tobiasraabe** [sphinx-doc/sphinx#7839]
**Describe the bug**

I want to use `autosummary` to document functions which contain German umlauts like ä, ö, ü. As an example, I use the following directive
```rst
.. autosummary::
   :toctree: _generated

   mypackage.module
```
The sub file is successfully generated in `docs/_generated/mypackage.module.rst`, but the encoding is `Western (Windows 1252)` which produces similar warnings during the build

```
reading sources... [100%] functions/_generated/gettsim.soz_vers.krankenv_pflegev
xxx\gettsim\docs\functions\_generated\gettsim.soz_vers.krankenv_pflegev.rst:20: WARNING: undecodable source characters, replacing with "?": b'      bezugsgr>>>\xf6<<<\xdfe\r'
xxx\gettsim\docs\functions\_generated\gettsim.soz_vers.krankenv_pflegev.rst:20: WARNING: undecodable source characters, replacing with "?": b'      bezugsgr\xf6>>>\xdf<<<e\r'
```

Converting the encoding of the file to `utf-8` and setting `autosummary_generate_overwrite = False`. Solves the issue.

**To Reproduce**

*Skipped*

**Expected behavior**

No warnings. Same output as here (https://gettsim--197.org.readthedocs.build/en/197/functions/_generated/gettsim.soz_vers.krankenv_pflegev.html#gettsim.soz_vers.krankenv_pflegev.bezugsgrosze). For example, function `bezugsgröße()`.

**Your project**

- [no error in rtd log because ubuntu](https://readthedocs.org/projects/gettsim/builds/11256848/)
- [the pr where the issue occurred](https://github.com/iza-institute-of-labor-economics/gettsim/pull/197)

**Environment info**
- OS: Windows
- Python version: 3.7.6
- Sphinx version: 3.1.1
- Sphinx extensions:  sphinx.ext.autosummary

**Comment by tk0miya** [comment]
The autosummary (and autodoc) read the docstring from python object. So they are loaded in correct character encodings by python interpreter. So you need to set the encoding of the code to your code via `coding` comment. Please ensure your source code has `coding:` comment.
https://docs.python.org/3/howto/unicode.html#unicode-literals-in-python-source-code
https://www.python.org/dev/peps/pep-0263/

**Comment by tobiasraabe** [comment]
I tried setting `# coding: utf-8` at the top of the source py file, but it does not change anything.

I do not think it has something to do with the source code. The source file is encoded with utf-8 and the generated stub files look fine at first glance. The umlauts are correctly displayed. But, the generated stub files have different encodings depending on whether Sphinx is executed on Windows or on Linux.

```
# Encoding of the source file
$ file ../gettsim/soz_vers/krankenv_pflegev.py
../gettsim/soz_vers/krankenv_pflegev.py: Python script, UTF-8 Unicode text executable

# Encoding of the stub file when generated on Ubuntu.
$ file _generated/gettsim.soz_vers.krankenv_pflegev.rst
_generated/gettsim.soz_vers.krankenv_pflegev.rst: UTF-8 Unicode text

# Encoding of the stub file when generated on Windows.
$ file _generated/gettsim.soz_vers.krankenv_pflegev.rst
_generated/gettsim.soz_vers.krankenv_pflegev.rst: ISO-8859 text, with CRLF line terminators
```
If I change the encoding from ISO-8859 to utf-8 manually and prevent that the file is regenerated, then, the documentation is also sucessfully built on Windows.

Thus, I suspect that Sphinx needs to change the encoding of the stub files it is generating.

If I missed your point, I am sorry.

**Comment by tobiasraabe** [comment]
And the stub file looks like this

```
# _generated/gettsim.soz_vers.krankenv_pflegev.rst

gettsim.soz\_vers.krankenv\_pflegev
===================================

.. automodule:: gettsim.soz_vers.krankenv_pflegev

   .. rubric:: Functions

   .. autosummary::
   
      ag_beitr_krankenv_midi_job
      ag_beitr_pflegev_midi_job
      an_beitr_krankenv_midi_job
      an_beitr_pflegev_midi_job
      bezugsgröße
      ges_beitr_krankenv_midi_job
      ges_beitr_pflegev_midi_job
      ges_krankenv_beitr_m
      ges_krankenv_beitr_m_tu
      ges_krankenv_beitr_rente
      ges_krankenv_beitr_selbst
      krankenv_beitr_bemess_grenze
      krankenv_beitr_regulär_beschäftigt
      krankenv_pflichtig_eink_selbst
      krankenv_pflichtig_rente
      lohn_krankenv__regulär_beschäftigt
      pflegev_beitr__regulär_beschäftigt
      pflegev_beitr_m
      pflegev_beitr_rente
      pflegev_beitr_selbst
      pflegev_zusatz_kinderlos
      selbsständig_ges_krankenv
```

**Comment by tk0miya** [comment]
Thank you for your explanation. Now I understand your problem in detail. Could you try #7843, please?

**Comment by tobiasraabe** [comment]
Thank your very much, @tk0miya :). I tested the fixed version and it passes the utf-8 encoding down to the stub file. The documentation can be generated sucessfully and without warnings.

However, the files have a slightly different encoding. The stub file is encoded with BOM. Is this an issue?

```
xxx/docs$ file _generated/gettsim.soz_vers.krankenv_pflegev.rst
_generated/gettsim.soz_vers.krankenv_pflegev.rst: UTF-8 Unicode (with BOM) text, with CRLF line terminators
xxx/docs$ file ../gettsim/soz_vers/krankenv_pflegev.py
../gettsim/soz_vers/krankenv_pflegev.py: Python script, UTF-8 Unicode text executable
```

**Comment by tk0miya** [comment]
#7843 uses `source_encoding` settings to the encoding of the reST file. It defaults to `utf-8_sig`. So please change it to `'utf-8'`.

**Comment by tk0miya** [comment]
Thank you for confirming. I just merged #7843 to 3.x branch. It will be released as the next release; 3.2.
